test/Microsoft.ML.Functional.Tests/Training.cs

```diff
@@ -0,0 +1,534 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Linq;
+using Microsoft.ML.Data;
+using Microsoft.ML.Functional.Tests.Datasets;
+using Microsoft.ML.RunTests;
+using Microsoft.ML.TestFramework;
+using Microsoft.ML.Trainers;
+using Microsoft.ML.Trainers.FastTree;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.ML.Functional.Tests
+{
+    public class Training : BaseTestClass
+    {
+        public Training(ITestOutputHelper output) : base(output)
+        {
+        }
+
+        /// <summary>
+        /// Training: It is easy to compare trainer evaluations on the same dataset.
+        /// </summary>
+        [Fact]
+        public void CompareTrainerEvaluations()
+        {
+            var mlContext = new MLContext(seed: 1);
+
+            // Get the dataset.
+            var data = mlContext.Data.LoadFromTextFile<TweetSentiment>(GetDataPath(TestDatasets.Sentiment.trainFilename),
+                separatorChar: TestDatasets.Sentiment.fileSeparator, hasHeader: TestDatasets.Sentiment.fileHasHeader, 
+                allowQuoting: TestDatasets.Sentiment.allowQuoting);
+            var trainTestSplit = mlContext.BinaryClassification.TrainTestSplit(data);
+            var trainData = trainTestSplit.TrainSet;
+            var testData = trainTestSplit.TestSet;
+
+            // Create a transformation pipeline.
+            var featurizationPipeline = mlContext.Transforms.Text.FeaturizeText("Features", "SentimentText")
+                .AppendCacheCheckpoint(mlContext);
+
+            // Create a selection of learners.
+            var sdcaTrainer = mlContext.BinaryClassification.Trainers.StochasticDualCoordinateAscent(
+                    new SdcaBinaryTrainer.Options { NumberOfThreads = 1 });
+
+            var fastTreeTrainer = mlContext.BinaryClassification.Trainers.FastTree(
+                    new FastTreeBinaryClassificationTrainer.Options { NumberOfThreads = 1 });
+
+            var ffmTrainer = mlContext.BinaryClassification.Trainers.FieldAwareFactorizationMachine(
+                    new FieldAwareFactorizationMachineBinaryClassificationTrainer.Options { });
```

Comment:
>new FieldAwareFactorizationMachineBinaryClassificationTrainer.Options { } [](start = 20, length = 73)

is this a required parameter? having to do this looks appalling :)

Reply:
Yes. We can't have an empty `()`. Filed #2927

---
In reply to: [264535384](https://github.com/dotnet/machinelearning/pull/2921#discussion_r264535384) [](ancestors = 264535384)